core/services/ardupilot_manager/firmware_upload/FirmwareUpload.py

```diff
@@ -0,0 +1,86 @@
+import pathlib
+import shutil
+import subprocess
+import threading
+
+from loguru import logger
+
+
+class FirmwareUploadFail(RuntimeError):
+    pass
+
+
+class UploadToolNotFound(RuntimeError):
+    pass
+
+
+class FirmwareUpload:
+    def __init__(self) -> None:
+        self._autopilot_port: pathlib.Path = pathlib.Path("/dev/autopilot")
+        self._baudrate_flightstack: int = 57600
+        self._baudrate_bootloader: int = 115200
+
+        binary_path = shutil.which(self.binary_name())
+        if binary_path is None:
+            logger.error("Uploader binary not found on system's PATH.")
+            raise UploadToolNotFound
+
+        self._binary = pathlib.Path(binary_path)
+        assert self.is_binary_working()
+
+    @staticmethod
+    def binary_name() -> str:
+        return "ardupilot_fw_uploader.py"
+
+    def binary(self) -> pathlib.Path:
+        return self._binary
+
+    def is_binary_working(self) -> bool:
+        if self.binary() is None:
+            return False
+
+        try:
+            subprocess.check_output([self.binary(), "--help"])
+            return True
+        except subprocess.CalledProcessError as error:
+            logger.exception(f"Binary returned {error.returncode} on '--help' call: {error.output}")
+            return False
+
+    def set_autopilot_port(self, port: pathlib.Path) -> None:
+        self._autopilot_port = port
+
+    def set_baudrate_flightstack(self, baudrate: int) -> None:
+        self._baudrate_flightstack = baudrate
+
+    def set_baudrate_bootloader(self, baudrate: int) -> None:
+        self._baudrate_bootloader = baudrate
+
+    def upload(self, firmware_path: pathlib.Path) -> None:
+        logger.info("Starting upload of firmware to board.")
+        # pylint: disable=consider-using-with
+        process = subprocess.Popen(
+            f"{self.binary()} {firmware_path}"
+            f" --baud-bootloader {self._baudrate_bootloader}"
+            f" --baud-flightstack {self._baudrate_flightstack}",
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            shell=True,
+            encoding="utf-8",
+        )
+
+        # Using a timer of 40 seconds to prevent being stuck on upload. Upload usually takes 20 seconds.
+        timer = threading.Timer(40, process.kill)
+        try:
+            timer.start()
+            for line in iter(process.stdout.readline, b""):
+                logger.debug(line)
```

Comment:
this will stream the logs to the terminal, but I'm having a hard time seeing how we'll show them to the user...

Reply:
To show it to the user it will need some changes. For now the API seems complete to me, in the sense that it tries to flash the firmware, it has a timeout, a proper response in both success and fail cases and it logs correctly.

Reply:
ok, this is a good start. I just checked and the assynchronous progress updates is tagged STABLE+, so we're good for now

Reply:
Cool! Haven't noticed that.